File: sdk/java/src/main/java/io/openkedge/atp/internal/Varint.java
package io.openkedge.atp.internal;

import java.io.ByteArrayOutputStream;

/**
 * LEB128 varints (minimal/canonical) and ZigZag, plus a bounds-checked reader.
 * ATP-0001 §6.2.
 *
 * <p>{@code uvarint} is unsigned LEB128 over an unsigned 64-bit value using the
 * shortest encoding (at most 10 bytes). Signed integers use ZigZag then uvarint.
 */
public final class Varint {
    private Varint() {}

    /** Unsigned LEB128, minimal form. {@code n} is treated as unsigned 64-bit. */
    public static byte[] uvarint(long n) {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        while (true) {
            int b = (int) (n & 0x7f);
            n >>>= 7;                    // unsigned shift
            if (n != 0) {
                out.write(b | 0x80);
            } else {
                out.write(b);
                break;
            }
        }
        return out.toByteArray();
    }

    /** ZigZag map: (n << 1) XOR (n >> 63), arithmetic shift for the sign. */
    public static long zigzag(long n) {
        return (n << 1) ^ (n >> 63);
    }

    public static long unzigzag(long u) {
        return (u >>> 1) ^ -(u & 1);
    }

    /** Signed varint: ZigZag followed by uvarint. */
    public static byte[] svarint(long n) {
        return uvarint(zigzag(n));
    }

    /** Non-canonical or truncated ATP encoding. */
    public static final class DecodeException extends Exception {
        private static final long serialVersionUID = 1L;

        public DecodeException() {
            super("non-canonical or truncated ATP encoding");
        }
    }

    /** A cursor over a byte array with bounds-checked, canonicality-enforcing reads. */
    public static final class Reader {
        private final byte[] buf;
        private int pos;

        public Reader(byte[] buf) {
            this.buf = buf;
            this.pos = 0;
        }

        public int position() {
            return pos;
        }

        public byte[] take(int n) throws DecodeException {
            if (n < 0) {
                throw new DecodeException();
            }
            long end = (long) pos + n;
            if (end > buf.length) {
                throw new DecodeException();
            }
            byte[] s = new byte[n];
            System.arraycopy(buf, pos, s, 0, n);
            pos = (int) end;
            return s;
        }

        public int u8() throws DecodeException {
            return take(1)[0] & 0xff;
        }

        /** Read a minimal uvarint; reject overflow and non-canonical encodings. */
        public long uvarint() throws DecodeException {
            int start = pos;
            long val = 0;
            int nbytes = 0;
            while (true) {
                int b = u8();
                // Guard against shift overflow past 64 bits.
                if (nbytes == 9 && (b & 0x7f) > 1) {
                    throw new DecodeException();
                }
                val |= (long) (b & 0x7f) << (7 * nbytes);
                nbytes++;
                if ((b & 0x80) == 0) {
                    break;
                }
                if (nbytes >= 10) {
                    throw new DecodeException();
                }
            }
            // Minimality: the canonical encoding must equal what we consumed.
            byte[] canonical = Varint.uvarint(val);
            if (canonical.length != pos - start) {
                throw new DecodeException();
            }
            for (int i = 0; i < canonical.length; i++) {
                if (canonical[i] != buf[start + i]) {
                    throw new DecodeException();
                }
            }
            return val;
        }

        public boolean eof() {
            return pos == buf.length;
        }
    }
}

File: reference/rust/src/varint.rs
//! LEB128 varints (minimal/canonical) and ZigZag, plus a bounds-checked reader.
//! ATP-0001 §6.2.

use std::fmt;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct DecodeError;

impl fmt::Display for DecodeError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str("non-canonical or truncated ATP encoding")
    }
}

impl std::error::Error for DecodeError {}

/// Unsigned LEB128, minimal form.
pub fn uvarint(mut n: u64) -> Vec<u8> {
    let mut out = Vec::new();
    loop {
        let b = (n & 0x7f) as u8;
        n >>= 7;
        if n != 0 {
            out.push(b | 0x80);
        } else {
            out.push(b);
            break;
        }
    }
    out
}

pub fn zigzag(n: i64) -> u64 {
    ((n << 1) ^ (n >> 63)) as u64
}

pub fn unzigzag(u: u64) -> i64 {
    ((u >> 1) as i64) ^ -((u & 1) as i64)
}

pub fn svarint(n: i64) -> Vec<u8> {
    uvarint(zigzag(n))
}

/// A cursor over a byte slice with bounds-checked reads.
pub struct Reader<'a> {
    pub buf: &'a [u8],
    pub pos: usize,
}

impl<'a> Reader<'a> {
    pub fn new(buf: &'a [u8]) -> Self {
        Reader { buf, pos: 0 }
    }

    pub fn take(&mut self, n: usize) -> Result<&'a [u8], DecodeError> {
        let end = self.pos.checked_add(n).ok_or(DecodeError)?;
        if end > self.buf.len() {
            return Err(DecodeError);
        }
        let s = &self.buf[self.pos..end];
        self.pos = end;
        Ok(s)
    }

    pub fn u8(&mut self) -> Result<u8, DecodeError> {
        Ok(self.take(1)?[0])
    }

    pub fn uvarint(&mut self) -> Result<u64, DecodeError> {
        let start = self.pos;
        let mut val: u128 = 0;
        let mut nbytes: u32 = 0;
        loop {
            let byte = self.u8()?;
            val |= ((byte & 0x7f) as u128) << (7 * nbytes);
            nbytes += 1;
            if byte & 0x80 == 0 {
                break;
            }
            if nbytes >= 10 {
                return Err(DecodeError);
            }
        }
        if val > u64::MAX as u128 {
            return Err(DecodeError);
        }
        let val = val as u64;
        // minimality: the canonical encoding must equal what we consumed.
        if uvarint(val) != self.buf[start..self.pos] {
            return Err(DecodeError);
        }
        Ok(val)
    }

    pub fn eof(&self) -> bool {
        self.pos == self.buf.len()
    }
}
